inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -737,6 +751,80 @@ void InsertCopyLayerPass::run() {
     }
 }
 
+void Concat4Dto2DPass::run() {
+    auto quantized = InferenceEngine::getInjectedData<QuantizedLayerParams>(pLayers->front());
+
+    if (getPassManager()->getPolicy().ConcatConversionPolicy == Policy::Concat4Dto2DConversion::DISABLED) return;
+    if (getPassManager()->getPolicy().ConcatAlignmentPolicy == Policy::ConcatAlignment::DISABLED) return;
+    if (getPassManager()->getPolicy().ConcatAlignmentPolicy == Policy::ConcatAlignment::DISABLED_FOR_FP32 && !quantized) return;
+
+    for (auto & l : *pLayers) {
+        LayerInfo info(l);
+        if (!info.isConcat()) continue;
+        auto concatLayer = info.as<ConcatLayer*>();
+
+        auto dims_size = concatLayer->insData[0].lock()->getDims().size();
+        if (dims_size > 2) {
+            auto axis = concatLayer->_axis;
+            bool skip_layer = false;
+            for (int i = 0; i < axis; i++) {
+                if (concatLayer->insData[0].lock()->getDims()[i] != 1) skip_layer = true;
```

Comment:
does it mean that this transformation works for cases when all dimensions are equal to 1? means only for trivial cases for first concat input?

Reply:
non axis layers must match by definition, thats why only first input is checked, and yes it covers only trivial cases since for non trivial cases 4D to 2D conversion may break the outcome (it is for cases where all dimensions before axis equal to 1, eq 1 5 3 2 input will accept axis 0 and 1 as correct since converting this type of concat to 2D will not modify the way the concat works, since it will be simple addition of data one input after another without reordering